Lemon: Vendingmachine step3

puuurm opened the pull request:
자판기 구조체 내부에 재고 상품 프로퍼티와 구매한 목록 프로퍼티를 둘 다 선언하였습니다. 비효율 적인거 아닌가 고민했는데 구매 목록을 불러올 때 구매 목록 프로퍼티가 없을 경우 더 복잡한 코드가 필요해서 둘 다 선언했습니다.

puuurm committed:
Modify README

puuurm committed:
Create VendingMachine swift File

puuurm committed:
Add properties and methods on VendingMachine

puuurm committed:
Move unit test string extension

puuurm committed:
Add VendingMachine unit test code

Fix bugs in the VendingMachine structure throught unit test.

puuurm committed:
Create InputView

puuurm committed:
Add methods on InputView

puuurm committed:
Create OutputView

puuurm committed:
Modify methods on VendingMachine

Modify unit test on VendingMachine.

puuurm committed:
Create methods on OutputView

puuurm committed:
Create invalidProductNumber error on VendingMachine

puuurm committed:
Create swift file for NSObject extension

puuurm committed:
Delete NSObjectExtension file and Modify property name

puuurm committed:
Add integrated test for vendingMachine

puuurm commented:
구조체 간의 관계 입니다.
[InputView] -> InputFormat(option: addMoney, buyProduct, detail: 돈의 액수, 구매 음료수 인덱스) -> [VendingMachine] -> [OutputView]

godrm commented on VendingMachine/VendingMachine/Drink/Drink.swift:
```diff
@@ -9,40 +9,54 @@
-    private(set) var price: String
+    private(set) var price: Int
     private(set) var name: String
     private(set) var dateOfManufacture: String
```
dateOfManufacture 는 Date 타입을 사용하도록 하세요
typeOfProduct는 init 매개변수로 받지말고 Class마다 자신의 타입을 지정하도록 구현하세요. 
weight도 Int로 하는게 좋을 것 같아요. 혹시 일부러 String 으로 하고 있나요?

puuurm replied:
굳이 연산하지 않는 경우는 String으로 뒀습니다. 다른 프로퍼티는 변경해도 상관 없는데, dateOfManufacturer경우는 String으로 초기화 해서 Date로 바꾸고, description에서 출력할 때 또 String으로 변경해줘야 하는데 그래도 Date 타입이 나을 까요?

puuurm replied:
typeOfProduct을 매개변수로 받지 않으면, 딸기우유, 바나나우유 이런 식으로 각각의 클래스에 스트링 값을 임의로 넣어주는 get 프로퍼티를 사용하는 건가요?

godrm replied:
String 타입으로 지정하는건 항상 마지막 옵션으로 선택하세요. 정확히 데이터구조를 표현하는 다른 타입이 있으면  그 타입을 활용하세요. 특히 날짜/시간을 표현하는건 캘린터/시간대/지역 표현과 독립적인 Date 타입으로 저장하는게 더 좋습니다. 출력할 때만 formatter를 활용하는게 좋습니다.

godrm replied:
typeOfProduct가 각 클래스별로 하나씩 지정가능하지 않은가요? 하위 클래스로 상속을 받을수록 세부 사항을 init 함수에서 값으로 지정할 수 있도록 설계하는게 좋습니다. 그렇지 않으면 하위 클래스로 갈수록 init 함수 매개변수가 줄어들지 않습니다.

godrm commented on VendingMachine/VendingMachine/Drink/SoftDrink.swift:
```diff
@@ -10,7 +10,15 @@ import Foundation
+        guard let manufacturingDate = dateFormatter.date(from: self.dateOfManufacture) else {
+            return nil
+        }
+        return Date(timeInterval: 3600 * 24 * 365 * 2, since: manufacturingDate)
```
expirationDate 도 유통기한 날짜만 다르다면 코드를 일반화 시킬 수 있지 않을까요?

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,128 @@
+    @discardableResult mutating func buy(productIndex: Int) throws -> Drink {
+        let listOfCanBuy = self.listOfCanBuy()
+        if productIndex < 1 || productIndex > listOfCanBuy.count {
+            throw stockError.invalidProductNumber
```
여기는 if 문이고 아래는 guard let 구문으로 throw를 하고 있네요. 두 경우는 왜 처리가 다른가요?

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,128 @@
+    }
+    // 전체 상품 재고를 (사전으로 표현하는) 종류별로 리턴하는 메소드
+    func listOfInventory() -> Array<(key: Drink, value: Count)> {
+        let sortedInventory = inventory.sorted(by: < )
```
이 함수는 요구사항이 `(사전으로 표현하는)` -> Dictionary를 의미합니다. Array<Dictionary<>> 형태가 아니라 inventory 처럼 Dictionary를 전달하세요.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+    // 시작이후 구매 상품 이력을 배열로 리턴하는 메소드
+    func listOfPurchase() -> Array<(key: DrinkName, value: Count)> {
```
이제 Drink 같은 클래스로 설계하기 시작하면, Drink 객체 인스턴스 자체를 배열에 넣는 형태로 직접 활용하는게 좋습니다. 
(당장 수정할 필요는 없어 보입니다)

puuurm committed:
Modify properties

Assign a default value to typeOfProduct.
Change dateOfManufacture to Date type. Create maintenanceDay and generalize expirationDate property.
Generate convertToFloat() for ratio extracion.

puuurm committed:
Modify type of purchases on VendingMachine

puuurm committed:
Add case of extractRemainMoney on VendingMachine

puuurm commented:
1. 피드백 반영하였습니다.
2. Input에서 음료수를 더이상 구매하지 않고 잔돈을 출금하는 extractRemainMoney 옵션을 추가하였습니다. 잔돈을 출금하면, 자판기에 남아있는 금액은 0이 되고, 구매 목록을 출력합니다.

puuurm committed:
Apply codacy

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -43,8 +40,41 @@ let coffee = Coffee(productTpye: "커피", calorie: "150kcal",
+while(true) {
+    do {
+        let input = try inputView.read()
+        switch input.option {
```
try - catch가 필요한 함수가 아닌 switch 구문 아래 부분은 do-try-catch 이후로 빼는게 좋습니다.

godrm commented on VendingMachine/VendingMachine/Drink/Drink.swift:
```diff
@@ -9,49 +9,100 @@
+        guard let expirationDay = self.expirationDate else {
+            return false
+        }
+        return date < expirationDay ? true : false
```
사소한 거지만 `return date < expirationDay ? true : false` 대신 `return (date < expirationDay)` 정도가 더 깔끔합니다.

godrm commented on VendingMachine/VendingMachine/Drink/Drink.swift:
```diff
@@ -9,49 +9,100 @@
+    formatter.dateFormat = "yyyyMMdd"
+    formatter.timeZone = TimeZone(abbreviation: "GMT+0:00")
+    return formatter
+}()
```
전역 변수 dateFormatter를 전역 변수가 아니라 모듈 전체에서 쓸 수 있는 방법을 고민해보세요.

puuurm committed:
Generate numberFormatter on OutputView

puuurm committed:
Change the global constant dateFormatter to a private constant

puuurm committed:
Modify methods for short code

puuurm committed:
Modify printListOfAllPurchases()

puuurm commented:
1. 피드백 반영하였습니다.
2. OutputView에 numberFormatter 추가하였습니다.
3. 구매 목록에 총 금액 출력을 추가했습니다.

godrm commented:
수고하셨습니다. 다음 단계에 도전해보세요 👍
